### PR TITLE
Async combinators

### LINKED ISSUE #244: async combinators for Stream
Our [stream combinators](https://docs.rs/async-std/0.99.7/async_std/stream/trait.Stream.html) operate on asynchronous data, but are themselves synchronous.

This means it's not possible to for example filter or map a stream based on other async operations. Which seems limiting, as it's possible to call sync functions from async functions but not the other way around.

I'd like to propose we update our Stream combinators that take closures to be async functions.

cc/ @montekki 
ref #129

### PR DESCRIPTION
**Pull request by montekki**
Partially fixes #244 (Probably that issue needs to be broken down in two: 1) for what can be done now 2) what can be done in the future like async closures and taking by ref.

Also unimplements `Unpin` for `all` and `any` combinators becasue now we have a future cached that may not be unpin.

### CODE REVIEW

**Comment by yoshuawuyts** [comment]
This is very exciting; thanks for putting in the work!

**Comment by montekki** [comment]
@yoshuawuyts can we r+ #263 so this one can be updated and move on from draft?

File: src/stream/stream/mod.rs [hunk 1]
```diff
@@ -484,7 +486,12 @@ extension_trait! {
 
             let s: VecDeque<&str> = vec!["1", "lol", "3", "NaN", "5"].into_iter().collect();
 
-            let mut parsed = s.filter_map(|a| a.parse::<u32>().ok());
+            let mut parsed = s.filter_map(|a|
+                async move {
+                    a.parse::<u32>().ok()
+                });
+
+            let mut parsed = unsafe { std::pin::Pin::new_unchecked(&mut parsed) };
```
Comment (yoshuawuyts) [inline src/stream/stream/mod.rs]: Hm is there something we could do about the `unsafe` block here? Even using `pin_utils` might be better than what we currently have, but I'm curious why we need it?

Comment (montekki) [reply src/stream/stream/mod.rs]: yep, that's must be some leftover

File: src/stream/stream/mod.rs [hunk 2]
```diff
@@ -797,27 +816,28 @@ extension_trait! {
...
-            let s: VecDeque<usize> = vec![1, 2, 3].into_iter().collect();
-            let sum = s.for_each(move |x| tx.clone().send(x).unwrap()).await;
+            let mut x = 0;
 
-            let v: Vec<_> = rx.iter().collect();
+            let s: VecDeque<_> = vec![1u8, 2, 3].into_iter().collect();
+            s.for_each(|item| {
+                x += item;
+                futures::future::ready(())
+            }).await;
```
Comment (yoshuawuyts) [inline src/stream/stream/mod.rs]: Could we perhaps use an async block here?

Comment (montekki) [reply src/stream/stream/mod.rs]: capturing in async closures doesn't look that well now, but I will cook an example with Arc and Mutex

**Comment by montekki** [comment]
@yoshuawuyts fixed, but the `for_each` example is now cringe. Idk how to dodge this kind of capturing trouble with the support from the language we have at hand r.n.

File: src/stream/stream/mod.rs (hunk 2 above)
Comment (yoshuawuyts) [reply src/stream/stream/mod.rs]: @montekki oh dang, yeah you're right -- I think what we had before is def better indeed. Perhaps we should revert it?

This is def yet another argument for having async closures!

Comment (montekki) [reply src/stream/stream/mod.rs]: done

**Comment by yoshuawuyts** [comment]
Dang, it seems that CI went and broke between the last commit and the most recent one. Currently trying to fix parts of it through #280, but I'm thinking I should probably just manually run the tests on this PR to confirm things are good and then merge (:

**Comment by yoshuawuyts** [comment]
@montekki could you try rebasing on master?

**Comment by yoshuawuyts** [comment]
@montekki could you try rebasing again?

**Comment by dignifiedquire** [comment]
@yoshuawuyts @montekki what is the state of this?

**Comment by montekki** [comment]
@dignifiedquire this still needed? merged with master, going to do required cleanups a bit later

**Comment by taiki-e** [comment]
As far as I know, this was postponed to the next major version: https://github.com/async-rs/async-std/pull/496#issuecomment-552377963 (perhaps because async closure is incomplete)

**Comment by montekki** [comment]
Ok no rush then.

**Comment by montekki** [comment]
Ok, at this point I think that I'd rather re-implement it from scratch when time comes than continue attempts to update this PR so closing.

**Pull request closed without merging**